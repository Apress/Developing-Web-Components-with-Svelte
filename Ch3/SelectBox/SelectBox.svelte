<svelte:options tag="cobalt-selectbox" />

<script>
  import selectOptionsData from "./selectboxdata.json";

  export let selectOptions = selectOptionsData;
  export let displayText = (a) => a.text;
  export let index = 0;
  export let disabled = false;

  function onChange(event) {
    event.target.dispatchEvent(new CustomEvent("onchange", { composed: true }));
  }
</script>

<div class="cobalt">
  <select bind:value={index} {disabled} on:change={onChange}>
    {#each selectOptions as option, i}
      <option value={i + 1}>{displayText(option)}</option>
    {/each}
  </select>
</div>

<style>
  .cobalt {
    display: flex;
  }

  select {
    padding: 5px 100px 5px 5px; /* 100px required to make sure image displays */
    font-size: 16px;
    border: 1px solid #19247c;
    height: 34px;
    border-radius: 10px;
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;
    background: url("./icon.png") 96% / 15% no-repeat #9393ff;
  }
</style>
