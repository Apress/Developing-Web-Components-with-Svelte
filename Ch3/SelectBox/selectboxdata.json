[
  { "text": "aaa" },
  { "text": "bbb" },
  { "text": "ccc" }
]