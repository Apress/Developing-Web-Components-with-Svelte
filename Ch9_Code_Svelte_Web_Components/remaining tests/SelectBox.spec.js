import { userEvent } from '@storybook/testing-library';
import { render, screen } from '@testing-library/svelte';
import SelectBox from "../src/lib/SelectBox/SelectBox.svelte";

describe("Tests for SelectBox", () => {
  const props = {
    selectOptions: [
      { "text": "aaa" },
      { "text": "bbb" },
      { "text": "ccc" }
    ],
    index: 0,
    disabled: false,
  };

  it("should render properly", () => {
    const result = render(SelectBox, props );
    expect(() => result).not.toThrow();
  });

  it("should select the second option if clicked", () => {
    const result = render(SelectBox, props );
    const selectedText = userEvent.selectOptions(screen.getAllByTestId("selectdropdown")[1], ["text"]);
    expect(selectedText).toContain("bbb");
  });

  it("get a snapshot of component", () => {
    const tree = render(SelectBox, props);
    expect(tree).toMatchSnapshot();
  })
});