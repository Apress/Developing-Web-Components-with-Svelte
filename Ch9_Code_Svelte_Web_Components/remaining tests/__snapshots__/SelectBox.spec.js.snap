// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for SelectBox get a snapshot of component 1`] = `
{
  "component": SelectBox {
    "$$": {
      "after_update": [],
      "before_update": [],
      "bound": {},
      "callbacks": {},
      "context": Map {},
      "ctx": [
        0,
        [
          {
            "text": "aaa",
          },
          {
            "text": "bbb",
          },
          {
            "text": "ccc",
          },
        ],
        [Function],
        false,
        [Function],
      ],
      "dirty": [
        -1,
      ],
      "fragment": {
        "c": [Function],
        "d": [Function],
        "i": [Function],
        "l": [Function],
        "m": [Function],
        "o": [Function],
        "p": [Function],
      },
      "not_equal": [Function],
      "on_destroy": [
        [Function],
      ],
      "on_disconnect": [],
      "on_mount": [],
      "props": {
        "disabled": 3,
        "displayText": 2,
        "index": 0,
        "selectOptions": 1,
      },
      "root": <div>
        <div
          class="cobalt svelte-13v4w3y"
        >
          <select
            class="svelte-13v4w3y"
          >
            <option
              data-testid="selectdropdown"
              value="1"
            >
              aaa
            </option>
            <option
              data-testid="selectdropdown"
              value="2"
            >
              bbb
            </option>
            <option
              data-testid="selectdropdown"
              value="3"
            >
              ccc
            </option>
          </select>
        </div>
      </div>,
      "skip_bound": false,
      "update": [Function],
    },
    "$$set": [Function],
    "$capture_state": [Function],
    "$inject_state": [Function],
  },
  "container": <body>
    <div>
      <div
        class="cobalt svelte-13v4w3y"
      >
        <select
          class="svelte-13v4w3y"
        >
          <option
            data-testid="selectdropdown"
            value="1"
          >
            aaa
          </option>
          <option
            data-testid="selectdropdown"
            value="2"
          >
            bbb
          </option>
          <option
            data-testid="selectdropdown"
            value="3"
          >
            ccc
          </option>
        </select>
      </div>
    </div>
  </body>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
